### PR TITLE
Support autoprobing floating-network-id for openstack cloud provider

### PR DESCRIPTION
**Pull request by FengyunPan**
Currently if user doesn't specify floatingnetwork-id and loadbalancer.openstack.org/floating-network-id annotation, openstack cloud provider can't create a external LoadBalancer service.
Actually we can get  floatingnetwork-id automatically. 
If we get multiple  floatingnetwork-ids, then ask user to specify one, or we use the  floatingnetwork-id to create floatingip for external LoadBalancer service.

This is a part of #50726

**Special notes for your reviewer**:
/assign @dims

**Release note**:
```release-note
Support autoprobing floating-network-id for openstack cloud provider
```

### CODE REVIEW

**Comment by dims** [comment]
@FengyunPan PR needs rebase

**Comment by xsgordon** [comment]
@FengyunPan @dims do we have a target milestone in mind for this?

**Comment by dims** [comment]
waiting for @anguslees :)

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 1]
```diff
@@ -560,6 +562,46 @@ func getNodeSecurityGroupIDForLB(compute *gophercloud.ServiceClient, nodes []*v1
...
+func getFloatingNetworkIdForLB(client *gophercloud.ServiceClient) (string, error) {
+	var FloatingNetworkIds []string
+
+	type NetworkWithExternalExt struct {
+		networks.Network
+		external.NetworkExternalExt
+	}
+	var externalNetwork []NetworkWithExternalExt
+	err := networks.List(client, networks.ListOpts{}).EachPage(func(page pagination.Page) (bool, error) {
+		err := networks.ExtractNetworksInto(page, &externalNetwork)
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Move the `externalNetwork` declaration 2 lines above to immediately above this line (_inside_ this pagination callback function).  That will help make sure nobody tries to use the value incorrectly outside/after this function.

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 2]
```diff
@@ -560,6 +562,46 @@ func getNodeSecurityGroupIDForLB(compute *gophercloud.ServiceClient, nodes []*v1
...
+		if err != nil {
+			return false, err
+		}
+
+		for _, externalNet := range externalNetwork {
+			if externalNet.External == true {
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Remove `== true`, `externalNet.External` is already a boolean.

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 3]
```diff
@@ -560,6 +562,46 @@ func getNodeSecurityGroupIDForLB(compute *gophercloud.ServiceClient, nodes []*v1
 	return nodeSecurityGroupIDs.List(), nil
 }
 
+// getFloatingNetworkIdForLB returns a floating-network-id for cluster.
+func getFloatingNetworkIdForLB(client *gophercloud.ServiceClient) (string, error) {
+	var FloatingNetworkIds []string
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: lowercase the first letter, following usual golang private/public conventions.

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 4]
```diff
@@ -560,6 +562,46 @@ func getNodeSecurityGroupIDForLB(compute *gophercloud.ServiceClient, nodes []*v1
...
+		}
+
+		for _, externalNet := range externalNetwork {
+			if externalNet.External == true {
+				FloatingNetworkIds = append(FloatingNetworkIds, externalNet.ID)
+			}
+		}
+
+		if len(FloatingNetworkIds) > 1 {
+			return false, ErrMultipleResults
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Hrm.  If we have multiple external networks, is it reasonable to just pick any one of them (eg the "first" one) - assuming we don't have more explicit configuration from the user?

Or if there is some sort of recognisable error when we get it wrong, perhaps we should try to use each of them in turn when setting up the LB?

  (trying to make this logic "just work" in as many cases as possible)

**Review by anguslees — CHANGES_REQUESTED** [review]

**Comment by anguslees** [comment]
(... Otherwise I :heart: this PR.  Less configuration is the best configuration)

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go (hunk 3 above)
Comment (FengyunPan) [reply pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Done.

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go (hunk 2 above)
Comment (FengyunPan) [reply pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Done.

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go (hunk 1 above)
Comment (FengyunPan) [reply pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Done.

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go (hunk 4 above)
Comment (FengyunPan) [reply pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Yeah, makes sense, if there are multiple external networks, I just return err, it is bad.
I update it soon.

**Comment by FengyunPan** [comment]
@anguslees Are these codes ok to approve?

**Comment by FengyunPan** [comment]
Just do a code rebase.

**Comment by FengyunPan** [comment]
@jdumars @calebamiles Can you please add the 1.9 milestone flag for this PR? Thank you.

**Comment by dims** [comment]
/lgtm
/approve

@enisoc @idvoretskyi @xsgordon Can we please get this into 1.9?

**Comment by jdumars** [comment]
Can someone from SIG OpenStack please acknowledge that this is approved for the 1.9 milestone?

**Comment by dims** [comment]
@jdumars Ack on behalf of SIG-openstack . yes please.

**Comment by enisoc** [comment]
@dims @FengyunPan This appears to be adding support rather than fixing something that's broken. Please file an [exception request](https://github.com/kubernetes/features/blob/master/EXCEPTIONS.md) if you need more time to get all the approvals.

**Comment by jdumars** [comment]
I overstepped on the escalation, sorry!  Anthony is correct, and this will need an exception.  Just make sure it gets +1's from the SIG on the email thread.  The intent is to ensure the SIG has oversight and outlines/accepts the risks of a late-breaking feature inclusion.

**Comment by dims** [comment]
/test pull-kubernetes-unit

@FengyunPan Can you please file an exception request?

**Comment by FengyunPan** [comment]
Just do a code rebase.
@dims Yes, I have filed an exception request to kubernetes-sig-openstack, please PTAL. (I do not know if it is ok.)

**Comment by dims** [comment]
/assign @thockin 

(for top level approval)

/approve

**Comment by jberkus** [comment]
Hey, can a SIG member please add this PR to the 1.9 milestone?  Thanks.

**Comment by jberkus** [comment]
changing priority so that it can stay in the milestone

/priority critical-urgent

**Comment by FengyunPan** [comment]
ping @thockin 
Please approve this PR, thank you.

**Comment by FengyunPan** [comment]
ping @lavalamp @smarterclayton @wojtek-t
For approval.

**Pull request merged**